cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -0,0 +1,1100 @@
+
+Status S3FileSystem::GetTargetStats(const std::string& s, FileStats* out) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+  FileStats st;
+  st.set_path(s);
+
+  if (path.empty()) {
+    // It's the root path ""
+    st.set_type(FileType::Directory);
+    *out = st;
+    return Status::OK();
+  } else if (path.key.empty()) {
+    // It's a bucket
+    S3Model::HeadBucketRequest req;
+    req.SetBucket(ToAwsString(path.bucket));
+
+    auto outcome = impl_->client_->HeadBucket(req);
+    if (!outcome.IsSuccess()) {
+      if (!IsNotFound(outcome.GetError())) {
+        return ErrorToStatus(outcome.GetError());
+      }
+      st.set_type(FileType::NonExistent);
+      *out = st;
+      return Status::OK();
+    }
+    // NOTE: S3 doesn't have a bucket modification time.  Only a creation
+    // time is available, and you have to list all buckets to get it.
+    st.set_type(FileType::Directory);
+    *out = st;
+    return Status::OK();
+  } else {
+    // It's an object
+    S3Model::HeadObjectRequest req;
+    req.SetBucket(ToAwsString(path.bucket));
+    req.SetKey(ToAwsString(path.key));
+
+    auto outcome = impl_->client_->HeadObject(req);
+    if (outcome.IsSuccess()) {
+      // "File" object found
+      *out = st;
+      return FileObjectToStats(outcome.GetResult(), out);
+    }
+    if (!IsNotFound(outcome.GetError())) {
+      return ErrorToStatus(outcome.GetError());
+    }
+    // Not found => perhaps it's an empty "directory"
+    bool is_dir = false;
+    RETURN_NOT_OK(impl_->IsEmptyDirectory(path, &is_dir));
+    if (is_dir) {
+      st.set_type(FileType::Directory);
+      *out = st;
+      return Status::OK();
+    }
+    // Not found => perhaps it's a non-empty "directory"
+    RETURN_NOT_OK(impl_->IsNonEmptyDirectory(path, &is_dir));
+    if (is_dir) {
+      st.set_type(FileType::Directory);
+    } else {
+      st.set_type(FileType::NonExistent);
+    }
+    *out = st;
+    return Status::OK();
+  }
+}
+
+Status S3FileSystem::GetTargetStats(const Selector& select, std::vector<FileStats>* out) {
+  S3Path base_path;
+  RETURN_NOT_OK(S3Path::FromString(select.base_dir, &base_path));
+  out->clear();
+
+  if (base_path.empty()) {
+    // List all buckets
+    std::vector<std::string> buckets;
+    RETURN_NOT_OK(impl_->ListBuckets(&buckets));
+    for (const auto& bucket : buckets) {
+      FileStats st;
+      st.set_path(bucket);
+      st.set_type(FileType::Directory);
+      out->push_back(std::move(st));
+      if (select.recursive) {
+        RETURN_NOT_OK(impl_->Walk(select, bucket, "", out));
+      }
+    }
+    return Status::OK();
+  }
+
+  // Nominal case -> walk a single bucket
+  return impl_->Walk(select, base_path.bucket, base_path.key, out);
+}
+
+Status S3FileSystem::CreateDir(const std::string& s, bool recursive) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+
+  if (path.key.empty()) {
+    // Create bucket
+    return impl_->CreateBucket(path.bucket);
```

Comment:
It seems like we should not make it _too_ easy to create buckets, but in AWS access controls, creating buckets may be disallowed (so this would simply fail). I don't think there's necessarily anything to change here